### PR TITLE
Comp. Svcs. SJC integration and test; added getMySubmissions() API

### PR DESCRIPTION
**Pull request by bkng**

### CODE REVIEW

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java [hunk 1]
```diff
@@ -2850,4 +2889,302 @@ public EntityIdList getDescendants(String nodeId, int generation, int pageSize,
...
+			throw new SynapseException(e);
+		}
+	}
+	
+	public Long getCompetitionCount() throws SynapseException {
+		PaginatedResults<Competition> res = getCompetitionsPaginated(0,0);
+		return res.getTotalNumberOfResults();
+	}
+	
+	public Competition findCompetition(String name) throws SynapseException {
```
Comment (brucehoff) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java:3027]: test with a name that requires URL encoding, e.g. a name with a space.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java [hunk 2]
```diff
@@ -2850,4 +2889,302 @@ public EntityIdList getDescendants(String nodeId, int generation, int pageSize,
...
+		}
+	}
+	
+	public void deleteCompetition(String compId) throws SynapseException {
+		if (compId == null) throw new IllegalArgumentException("Competition id cannot be null");
+		String uri = createEntityUri(COMPETITION_URI_PATH, compId);
+		deleteUri(uri);
+	}
+	
+	public Participant createParticipant(String compId) throws SynapseException {
```
Comment (brucehoff) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java:3062]: add a comment explaining that this adds the authenticated user as a participant in the given competition (evaluation)

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java [hunk 3]
```diff
@@ -2850,4 +2889,302 @@ public EntityIdList getDescendants(String nodeId, int generation, int pageSize,
...
+		if (compId == null) throw new IllegalArgumentException("Competition id cannot be null");
+		String uri = createEntityUri(COMPETITION_URI_PATH, compId) + "/" + PARTICIPANT;
+		JSONObject jsonObj = postUri(uri);
+		return initializeFromJSONObject(jsonObj, Participant.class);
+	}
+	
+	public Participant createParticipantAsAdmin(String compId, String idToAdd) throws SynapseException {
```
Comment (brucehoff) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java:3069]: idToAdd -> participantPrincipalId

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java [hunk 4]
```diff
@@ -2850,4 +2889,302 @@ public EntityIdList getDescendants(String nodeId, int generation, int pageSize,
...
+				+ "/" + principalId;		
+		JSONObject jsonObj = getEntity(uri);
+		JSONObjectAdapter adapter = new JSONObjectAdapterImpl(jsonObj);
+		try {
+			return new Participant(adapter);
+		} catch (JSONObjectAdapterException e1) {
+			throw new RuntimeException(e1);
+		}
+	}
+
```
Comment (brucehoff) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java:3007]: add comment "removes user/principal from a competition"

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java [hunk 5]
```diff
@@ -2850,4 +2889,302 @@ public EntityIdList getDescendants(String nodeId, int generation, int pageSize,
...
+			results.initializeFromJSONObject(adapter);
+			return results;
+		} catch (JSONObjectAdapterException e) {
+			throw new SynapseException(e);
+		}
+	}	
+	
+	public Long getSubmissionCount(String compId) throws SynapseException {
+		if (compId == null) throw new IllegalArgumentException("Competition id cannot be null");
+		PaginatedResults<Submission> res = getAllSubmissions(compId, 1, 0);
```
Comment (bkng) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java:3279]: should be limit 0 offset 0

File: integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+		dataset = synapseOne.createEntity(dataset);
+		projectTwo = synapseTwo.createEntity(new Project());
+		
+		entitiesToDelete.add(project.getId());
+		entitiesToDelete.add(dataset.getId());
+		entitiesToDelete.add(projectTwo.getId());
+		
+		// initialize Competitions
+		comp1 = new Competition();
+		comp1.setName("name");
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java:98]: use a name that is not url safe, e.g. containing a space

File: integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+		fetched = synapseOne.findCompetition(comp1.getName());
+		assertEquals(comp1, fetched);
+		
+		// Update
+		fetched.setDescription(comp1.getDescription() + " (modified)");
+		Competition updated = synapseOne.updateCompetition(fetched);
+		assertFalse("eTag was not updated", updated.getEtag().equals(fetched.getEtag()));
+		fetched.setEtag(updated.getEtag());
+		assertEquals(fetched, updated);
+		Long newCount = initialCount + 1;
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java:170]: move this to the 'create' section above

File: integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+		assertFalse("Etag was not updated", oldEtag.equals(comp1.getEtag()));
+		
+		String entityId1 = project.getId();
+		assertNotNull(entityId1);
+		entitiesToDelete.add(entityId1);
+		String entityId2 = dataset.getId();
+		assertNotNull(entityId2);
+		entitiesToDelete.add(entityId2);
+		
+		sub1.setCompetitionId(comp1.getId());
```
Comment (bkng) [inline integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java:300]: remove these submissions

File: integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+		sub2.setVersionNumber(1L);
+		sub2.setUserId(userName);
+		sub2 = synapseOne.createSubmission(sub2);
+		assertNotNull(sub2.getId());
+		submissionsToDelete.add(sub2.getId());
+				
+		// paginated submissions and bundles
+		PaginatedResults<Submission> subs;
+		PaginatedResults<SubmissionBundle> subBundles;
+
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT520SynapseJavaClientCompetitionTest.java:376]: please make sure that somewhere (manager level or here) we test that the participant cannot retrieve all submissions/bundles

File: lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/competition/model/SubmissionBundle.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+import org.sagebionetworks.schema.adapter.JSONObjectAdapter;
+import org.sagebionetworks.schema.adapter.JSONObjectAdapterException;
+
+/**
+ * Bundle to transport a Submission and its accompanying SubmissionStatus object
+ * 
+ * @author bkng
+ *
+ */
+public class SubmissionBundle implements JSONEntity {
```
Comment (brucehoff) [inline lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/competition/model/SubmissionBundle.java:15]: refactor as a json schema

File: lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/competition/model/SubmissionBundle.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+	
+	public static final String JSON_SUBMISSION = "submisison";
+	public static final String JSON_SUBMISSION_STATUS = "submissionStatus";
```
Comment (brucehoff) [inline lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/competition/model/SubmissionBundle.java:18]: submisison -> submission

**Comment by brucehoff** [comment]
great job!

**Pull request merged**